Tests/Opc.Ua.Core.Tests/Security/Certificates/CertificateValidatorTest.cs

```diff
@@ -652,29 +652,91 @@ public void VerifyPemWriterPrivateKeys()
             // all app certs are trusted
             foreach (var appCert in m_appSelfSignedCerts)
             {
-                var pemDataBlob = CertificateFactory.ExportPrivateKeyAsPEM(appCert);
+                var pemDataBlob = PEMWriter.ExportPrivateKeyAsPEM(appCert);
+                var pemString = Encoding.UTF8.GetString(pemDataBlob);
+                TestContext.Out.WriteLine(pemString);
+                CertificateFactory.CreateCertificateWithPEMPrivateKey(new X509Certificate2(appCert.RawData), pemDataBlob);
+                // note: password is ignored
+                var newCert = CertificateFactory.CreateCertificateWithPEMPrivateKey(new X509Certificate2(appCert.RawData), pemDataBlob, "password");
+                X509Utils.VerifyRSAKeyPair(newCert, newCert, true);
+            }
+        }
+
+        /// <summary>
+        /// Verify the PEM Writer, no password
+        /// </summary>
+        [Test, Order(502)]
+        public void VerifyPemWriterPublicKeys()
+        {
+            // all app certs are trusted
+            foreach (var appCert in m_appSelfSignedCerts)
+            {
+                var pemDataBlob = PEMWriter.ExportCertificateAsPEM(appCert);
+                var pemString = Encoding.UTF8.GetString(pemDataBlob);
+                TestContext.Out.WriteLine(pemString);
+#if NETCOREAPP3_1
+                var exception = Assert.Throws<ArgumentException>(() => { CertificateFactory.CreateCertificateWithPEMPrivateKey(new X509Certificate2(appCert), pemDataBlob); });
+#endif
+            }
+        }
+
+#if NETCOREAPP3_1
+        /// <summary>
+        /// Verify the PEM Writer, no password
+        /// </summary>
+        [Test, Order(501)]
+        public void VerifyPemWriterRSAPrivateKeys()
+        {
+            // all app certs are trusted
+            foreach (var appCert in m_appSelfSignedCerts)
+            {
+                var pemDataBlob = PEMWriter.ExportRSAPrivateKeyAsPEM(appCert);
                 var pemString = Encoding.UTF8.GetString(pemDataBlob);
-                CertificateFactory.CreateCertificateWithPEMPrivateKey(new X509Certificate2(appCert), pemDataBlob);
-                CertificateFactory.CreateCertificateWithPEMPrivateKey(new X509Certificate2(appCert), pemDataBlob, "password");
+                TestContext.Out.WriteLine(pemString);
+                var cert = CertificateFactory.CreateCertificateWithPEMPrivateKey(new X509Certificate2(appCert.RawData), pemDataBlob);
+                Assert.NotNull(cert);
+                // note: password is ignored
+                var newCert = CertificateFactory.CreateCertificateWithPEMPrivateKey(new X509Certificate2(appCert.RawData), pemDataBlob, "password");
+                X509Utils.VerifyRSAKeyPair(newCert, newCert, true);
             }
         }
 
+        /// <summary>
+        /// Verify the PEM Writer, with password
+        /// </summary>
+        [Test, Order(501)]
```

Comment:
Avoid using same order for different tests

Reply:
catch